plugins/uast-kotlin-base/src/org/jetbrains/uast/kotlin/KotlinAbstractUElement.kt

```diff
@@ -5,43 +5,39 @@
 
 package org.jetbrains.uast.kotlin
 
-import com.intellij.psi.PsiElement
-import org.jetbrains.uast.UElement
-import org.jetbrains.uast.UExpression
+import com.intellij.openapi.components.ServiceManager
+import org.jetbrains.uast.*
 import org.jetbrains.uast.kotlin.internal.KotlinUElementWithComments
 
-abstract class FirKotlinAbstractUElement(
-    private val givenParent: UElement?
+abstract class KotlinAbstractUElement(
+    givenParent: UElement?
 ) : KotlinUElementWithComments {
 
+    protected val languagePlugin: UastLanguagePlugin? by lz {
+        psi?.let { UastFacade.findPlugin(it) }
+    }
+
+    protected val baseResolveProviderService: BaseKotlinUastResolveProviderService? by lz {
```

Comment:
Is it ok that baseResolveProviderService may be null? Maybe we may always pass it as constructor parameter?

Reply:
Fortunately(?), parent conversions can return `null` (if something goes wrong, and the lack/failure of service/retrieval could be one reason). Passing it as constructor parameter is doable, but would be fairly large refactoring, so I'll wait for @xiexed input.

A couple more things to note: as mentioned in the PR description, I think plugin is a better fit for parent "convert" APIs. Also, only `UFile` has a plugin as a constructor parameter. All other declarations retrieve a plugin via `getLanguagePlugin`, which is deprecated as it could throw an exception unexpectedly. Instead, `findPlugin`, which may return `null` if a plugin is not available for some reasons, is preferred. I use that throughout FIR UAST, and I think this service retrieval is in a similar situation. So, this may-be-null resolve service seems consistent to me.

Reply:
I think that the lack of resolve service is the serious configuration error so we should not support this case and throw a exception instead of handling nullability. `getLanguagePlugin` is nullable because there could be languages without UAST plugin - it is ok, but the strate when we don't have a resolve service looks like a critical error

Reply:
Good to learn why language plugin could be nullable! Yeah, after knowing that, it makes sense to regard the lack of resolve provider service as a serious configuration error.

So, I tried to enforce the existence of it, and surprisingly (or not?), there is a hole where sources of service lookup, including psi, given parent, or declaration-specific elements, aren't available at all: declaration expression. For that case, constructor parameter in the form of an optional supplier seems good enough to me. See the latest commit.